src/lib/server/textGeneration/tools.ts

```diff
@@ -0,0 +1,187 @@
+	if (!tool) return { call, status: ToolResultStatus.Error, message: "Could not find tool" };
+
+	// Special case for directly_answer tool where we ignore
+	if (toolHasName("directly_answer", tool)) return;
+
+	yield {
+		type: MessageUpdateType.Tool,
+		subtype: MessageToolUpdateType.Call,
+		uuid,
+		call,
+	};
+	try {
+		const toolResult = yield* tool.call(call.parameters, {
+			conv,
+			messages,
+			preprompt,
+			assistant,
+		});
+		yield {
+			type: MessageUpdateType.Tool,
+			subtype: MessageToolUpdateType.Result,
+			uuid,
+			result: { ...toolResult, call } as ToolResult,
+		};
+		return { ...toolResult, call } as ToolResult;
+	} catch (cause) {
+		console.error(Error(`Failed while running tool ${call.name}`), { cause });
+		return {
+			call,
+			status: ToolResultStatus.Error,
+			message: cause instanceof Error ? cause.message : String(cause),
+		};
+	}
+}
+
+export async function* runTools(
+	{ endpoint, conv, messages, assistant }: TextGenerationContext,
+	tools: BackendTool[],
+	preprompt?: string
+): AsyncGenerator<MessageUpdate, ToolResult[], undefined> {
+	const calls: ToolCall[] = [];
+
+	// inform the model if there are files attached
+	const userMessage = messages.find((message) => message.from === "user");
+	preprompt = `${preprompt ?? ""}\n${makeFilesPrompt(userMessage?.files ?? [])}`.trim();
+
+	// do the function calling bits here
+	for await (const output of await endpoint({
+		messages,
+		preprompt,
+		generateSettings: assistant?.generateSettings,
+		tools,
+	})) {
+		// model natively supports tool calls
+		if (output.token.toolCalls) {
+			calls.push(...output.token.toolCalls);
+			continue;
+		}
+
+		// look for a code blocks of ```json and parse them
+		// if they're valid json, add them to the calls array
+		if (output.generated_text) {
+			const codeBlocks = Array.from(output.generated_text.matchAll(/```json\n(.*?)```/gs));
+			if (codeBlocks.length === 0) continue;
+
+			// grab only the capture group from the regex match
+			for (const [, block] of codeBlocks) {
+				try {
+					calls.push(
+						...JSON5.parse(block).filter(isExternalToolCall).map(externalToToolCall).filter(Boolean)
+					);
+				} catch (cause) {
+					// error parsing the calls
+					yield {
+						type: MessageUpdateType.Status,
+						status: MessageUpdateStatus.Error,
+						message: cause instanceof Error ? cause.message : String(cause),
+					};
+				}
+			}
+		}
+	}
+
+	const toolContext: BackendToolContext = { conv, messages, preprompt, assistant };
+	const toolResults: (ToolResult | undefined)[] = yield* mergeAsyncGenerators(
+		calls.map((call) => runTool(toolContext, tools, call))
+	);
+	return toolResults.filter((result): result is ToolResult => result !== undefined);
+}
+
+const externalToolCall = z.object({
+	tool_name: z.string(),
+	parameters: z.record(z.any()),
+});
+type ExternalToolCall = z.infer<typeof externalToolCall>;
+function isExternalToolCall(call: unknown): call is ExternalToolCall {
```

Comment:
```suggestion
});

type ExternalToolCall = z.infer<typeof externalToolCall>;

function isExternalToolCall(call: unknown): call is ExternalToolCall {
```

need some newlines for readability